docs/tutorials/essentials/part-4-using-data.md

```diff
@@ -441,6 +441,150 @@ const handleSubmit = (e: React.FormEvent<AddPostFormElements>) => {
+
+export const EditPostForm = () => {
+  const { postId } = useParams()
+
+  // highlight-next-line
+  const post = useAppSelector(state => selectPostById(state, postId!))
+  // omit component logic
+}
+```
+
+Note that the `postId` we get from `useParams()` is typed as `string | undefined`, but `selectPostById` expects a valid `string` as the argument. We can use the TS `!` operator to tell the TS compiler this value will not be `undefined` at this point in the code. (This can be dangerous, but we can make the assumption because we know the routing setup only shows `<EditPostForm>` if there's a post ID in the URL.)
+
+We'll continue this pattern of writing selectors in slices as we go forward, rather than writing them inline inside of `useAppSelector` in components. Remember, this isn't required, but it's a good pattern to follow!
+
+### Using Selectors Effectively
+
+It's often a good idea to encapsulate data lookups by writing reusable selectors. Ideally, components don't even have to know where in the Redux `state` a value lives - they just use a selector from the slice to access the data.
+
+You can also create "memoized" selectors that can help improve performance by optimizing rerenders and skipping unnecessary recalculations, which we'll look at in a later part of this tutorial.
+
+But, like any abstraction, it's not something you should do _all_ the time, everywhere. Writing selectors means more code to understand and maintain. **Don't feel like you need to write selectors for every single field of your state**. Try starting without any selectors, and add some later when you find yourself looking up the same values in many parts of your application code.
+
+### Optional: Defining Selectors Inside of `createSlice`
+
+We've seen that we can write selectors as standalone functions in slice files. In some cases, you can shorten this a bit by defining selectors directly inside `createSlice` itself.
+
+<DetailedExplanation title="Defining Selectors inside createSlice" >
+
+We've already seen that `createSlice` requires the `name`, `initialState`, and `reducers` fields, and also accepts an optional `extraReducers` field.
+
+If you want to define selectors directly inside of `createSlice`, you can pass in an additional `selectors` field. The `selectors` field should be an object similar to `reducers`, where the keys will be the selector function names, and the values are the selector functions to be generated.
+
+**Note that unlike writing a standalone selector function, the `state` argument to these selectors will be just the _slice state_, and _not_ the entire `RootState`!**.
+
+Here's what it might look like to convert the posts slice selectors to be defined inside of `createSlice`:
+
+```ts
+const postsSlice = createSlice({
+  name: 'posts',
+  initialState,
+  reducers: {
+    /* omit reducer logic */
+  },
+  // highlight-start
+  selectors: {
+    // Note that `state` here is just the `PostsState`!
+    selectAllPosts: state => state,
```

Comment:
I think it would be clearer for people if the variable here was named `postsState` instead of `state`.

People have associated the name `state` in selectors with `rootState` now. Calling it just "`state`" hear makes it confusing since it is the same name for a different concept.